[ARM][MVE] Add shuffle costs for LDn and STn instructions.

davemgreen opened the pull request:
LD2 is represented in IR as deinterleave-shuffle(load), and ST2 as store(interleave-shuffle). Whilst the shuffle would be expensive in general for MVE (it does not have zip/uzp instructions), it should be treated as cheap when part of the LD2/ST2 pattern. This borrows some code from the AArch64 backed to produce lower costs. (Some of which still shows as higher than it should - that just shows how broken the generic shuffle costs are at the moment, they would be lower if getShuffleCost was called directly as opposed to going through getInstructionCost).

john-brawn-arm commented on llvm/lib/Target/ARM/ARMTargetTransformInfo.cpp:
```diff
@@ -1336,6 +1336,37 @@ InstructionCost ARMTTIImpl::getShuffleCost(TTI::ShuffleKind Kind,
+           LT.second.getScalarSizeInBits() == 16 ||
+           LT.second.getScalarSizeInBits() == 32) &&
+          LT.second.getSizeInBits() == 128 &&
+          (ShuffleVectorInst::isDeInterleaveMaskOfFactor(Mask, 2) ||
```
Check MaxSupportedInterleaveFactor() >= 2 here? (So -mve-max-interleave-factor=1 will work as expected)

john-brawn-arm commented on llvm/lib/Target/ARM/ARMTargetTransformInfo.cpp:
```diff
@@ -1336,6 +1336,37 @@ InstructionCost ARMTTIImpl::getShuffleCost(TTI::ShuffleKind Kind,
+           LT.second.getScalarSizeInBits() == 32) &&
+          LT.second.getSizeInBits() == 128 &&
+          (ShuffleVectorInst::isDeInterleaveMaskOfFactor(Mask, 2) ||
+           (TLI->getMaxSupportedInterleaveFactor() == 4 &&
```
Should be >= 4, I think.

davemgreen replied:
It can only be 2 or 4 for MVE.

john-brawn-arm reviewed (commented):
There's a lot of duplication between the LD and ST code (I've added comments to the LD code but they also apply to the ST code). I think it makes sense to move at least of the logic here into a separate function.

davemgreen committed:
[ARM][MVE] Add shuffle costs for LDn and STn instructions.

LD2 is represented in IR as deinterleave-shuffle(load), and ST2 as
store(interleave-shuffle). Whilst the shuffle would be expensive in general for
MVE (it doesn not have zip/uzp instructions), it should be treated as cheap
when part of the LD2/ST2 pattern. This borrows some code from the AArch64
backed to produce lower costs. (Some of which still shows as higher than it
should - that just shows how broken the generic shuffle costs are at the
moment, they would be lower if getShuffleCost was called directly as opposed to
going through getInstructionCost).
